Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -0,0 +1,226 @@
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "DeferredWorkTimer.h"
+#include "JSPromise.h"
+#include <wtf/Condition.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/SentinelLinkedList.h>
+
+namespace JSC {
+
+enum class AtomicsWaitType : uint8_t { Sync, Async };
+enum class AtomicsWaitValidation : uint8_t { Pass, Fail };
+
+class Waiter final : public WTF::BasicRawSentinelNode<Waiter>, public ThreadSafeRefCounted<Waiter> {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    Waiter(VM* vm)
+        : m_isAsync(false)
+    {
+        m_vm = vm;
+    }
+
+    Waiter(JSPromise* promise)
+        : m_isAsync(true)
+    {
+        VM& vm = promise->vm();
+        {
+            JSLockHolder locker(vm);
+            m_ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+        }
+    }
+
+    bool isAsync()
+    {
+        return m_isAsync;
+    }
+
+    VM* getVM()
+    {
+        return m_isAsync ? &jsCast<JSPromise*>(m_ticket->target())->vm() : m_vm;
+    }
+
+    void nullOutVM()
+    {
+        ASSERT(!m_isAsync);
+        m_vm = nullptr;
+    }
+
+    Condition& getCondition()
+    {
+        ASSERT(!m_isAsync);
+        return m_condition;
+    }
+
+    DeferredWorkTimer::Ticket getTicket()
+    {
+        ASSERT(m_isAsync);
+        return m_ticket;
+    }
+
+    void setTimer(Ref<RunLoop::DispatchTimer>&& timer)
+    {
+        ASSERT(m_isAsync);
+        m_timer = WTFMove(timer);
+    }
+
+    bool hasTimer()
+    {
+        return !!m_timer;
+    }
+
+    void cancelTimer()
+    {
+        ASSERT(m_isAsync);
+        // If the timeout for AsyncWaiter is infinity, we won't dispatch any timer.
+        if (!m_timer)
+            return;
+        m_timer->stop();
+        m_timer->deref();
+    }
+
+private:
+    VM* m_vm { nullptr };
+    DeferredWorkTimer::Ticket m_ticket { nullptr };
+    RefPtr<RunLoop::DispatchTimer> m_timer { nullptr };
+    Condition m_condition;
+    bool m_isAsync { false };
+};
+
+class WaiterList {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    ~WaiterList()
+    {
+        removeIf([](Waiter*) {
+            return true;
+        });
+    }
+
+    void addLast(const AbstractLocker&, Waiter* waiter)
+    {
+        m_waiters.append(waiter);
+        // Since SyncWaiter will be allocated on stack, we only need to ref AsyncWaiter.
+        if (waiter->isAsync())
+            waiter->ref();
+        m_size++;
+    }
+
+    Waiter* takeFirst(const AbstractLocker&)
+    {
+        return take(&*m_waiters.begin());
+    }
```

Comment:
This looks wrong. Who is maintaining the ownership of the returned Waiter? It seems that it is releasing memory before returning.